File: Round 737/C.py
mod = 10**9+7
powers = [1]
for i in range(1, 200000+1):
	powers.append((powers[-1]*2)%mod)
for nt in range(int(input())):
	n, k = map(int,input().split())
	add = [1]
	for i in range(k):
		add.append((add[-1]*powers[n])%mod)
	if n%2:
		ans = pow(2, n*k, mod)
		for i in range(k):
			ans -= ((((pow(powers[n-1]+1, i, mod)) * (powers[n-1]-1))%mod) * add[k-i-1])%mod
			ans = ans%mod
		print (ans)
	else:
		ans = pow(2, n*k, mod)
		for i in range(k):
			ans -= ((((pow(powers[n-1]-1, i, mod)) * (powers[n-1]))%mod) * add[k-i-1])%mod
			ans = ans%mod
			# print (k, ans)
		print (ans)

File: Round 737/C.cpp
#include <iostream>
using namespace std;

long powers[200001];
long mod = 1000000007;

int pow(long long x, unsigned int y, int p)
{
	int res = 1;     // Initialize result
 
	x = x % p; // Update x if it is more than or
				// equal to p
  
	if (x == 0) return 0; // In case x is divisible by p;
 
	while (y > 0)
	{
		// If y is odd, multiply x with result
		if (y & 1)
			res = (res*x) % p;
 
		// y must be even now
		y = y>>1; // y = y/2
		x = (x*x) % p;
	}
	return res;
}

int main(){
	long nt, n, k;
	cin>>nt;
	powers[0] = 1;
	for(int i=1;i++;i<200001){
		powers[i] = (powers[i-1]*2)%mod;
	}
	while(nt--){
		cin>>n>>k;
		if (n%2==1){
			long ans = pow(2, n*k, mod);
			for (int i=0;i<k;i++){
				ans -= ((((pow(powers[n-1]+1, i, mod)) * (powers[n-1]-1))%mod) * pow(2, n*(k-i-1), mod))%mod;
				ans = ans%mod;
			}
			cout<<ans<<'\n';
		}else{
			long ans = pow(2, n*k, mod);
			for (int i=0;i<k;i++){
				ans -= ((((pow(powers[n-1]-1, i, mod)) * (powers[n-1]))%mod) * pow(2, n*(k-i-1), mod))%mod;
				ans = ans%mod;
			}
			cout<<ans<<'\n';
		}
	}
}